fendermint/testing/contract-test/tests/upgrades.rs

```diff
@@ -0,0 +1,300 @@
+        let mut state_params = self.state_params.clone();
+        state_params.timestamp = Timestamp(block_height as u64);
+
+        let state = FvmExecState::new(db, self.multi_engine.as_ref(), block_height, state_params)
+            .context("error creating new state")?
+            .with_block_hash(block_hash);
+
+        self.put_exec_state(state).await;
+
+        let res = self
+            .modify_exec_state(|s| self.interpreter.begin(s))
+            .await
+            .unwrap();
+        println!("Ret begin apply_ret : {:?}", res.apply_ret);
+
+        Ok(())
+    }
+
+    async fn end_block(&self, _block_height: ChainEpoch) -> Result<()> {
+        let _ret = self
+            .modify_exec_state(|s| self.interpreter.end(s))
+            .await
+            .context("end failed")?;
+
+        Ok(())
+    }
+
+    async fn commit(&mut self) -> Result<()> {
+        let exec_state = self.take_exec_state().await;
+
+        let (
+            state_root,
+            FvmUpdatableParams {
+                power_scale,
+                circ_supply,
+            },
+            _,
+        ) = exec_state.commit().context("failed to commit FVM")?;
+
+        self.state_params.state_root = state_root;
+        self.state_params.power_scale = power_scale;
+        self.state_params.circ_supply = circ_supply;
+
+        Ok(())
+    }
+}
+
+#[test]
+fn testest() {
+    let rt: tokio::runtime::Runtime =
+        tokio::runtime::Runtime::new().expect("create tokio runtime for init");
+
+    let mut us = UpgradeScheduler::default();
+
+    let chain_id = 1942764459484029;
+
+    us.add(Upgrade {
+        chain_id,
+        block_height: 1,
+        migration: |state| {
+            println!(
+                "Migration at height {} just prints out chainmetadata actor",
+                state.block_height()
+            );
+
+            let actor = state
+                .state_tree_mut()
+                .get_actor(system::SYSTEM_ACTOR_ID)
+                .unwrap();
+            print!("chainmetadata actor: {:?}", actor);
+            Ok(())
+        },
+    })
+    .unwrap();
+
+    us.add(Upgrade {
+        chain_id,
+        block_height: 2,
+        migration: |state| {
+            println!(
+                "Migration at height {} deletes chainmetadata actor",
+                state.block_height()
+            );
+
+            state.state_tree_mut().delete_actor(system::SYSTEM_ACTOR_ID);
+
+            Ok(())
+        },
+    })
+    .unwrap();
+
+    us.add(Upgrade {
+        chain_id,
+        block_height: 3,
+        migration: |state| {
+            println!(
+                "Migration at height {} confirms the chainmetadata actor is deleted",
+                state.block_height()
+            );
+
+            let actor = state
+                .state_tree_mut()
+                .get_actor(system::SYSTEM_ACTOR_ID)
+                .unwrap();
+            print!("chainmetadata actor after delete: {:?}", actor);
```

Comment:
Wrong ID. Maybe it could print `actor.is_none()`, it would be a bit more suggestive of what it expects.

Reply:
I'd actually enable these as assertions, so you can really test your expectations, and also some kind of assertion in the test that the upgrades _did_ run. Maybe you actually matched the chain ID to the "test"; without a comment it looked just like a random number. If someone changed it, it would be important to test that they didn't just skip all the upgrades. The app version could help with this.